### PR TITLE
Change to absolute pathing

### PR DESCRIPTION
**Pull request by Sieabah**
Relative pathing exists where it makes sense, internal to each module.

There are bugs dealing with Eua and User related services/classes/components that cannot be loaded with absolutes. I haven't had a chance to look further into it.

### CODE REVIEW

**Comment by ymao1** [comment]
Have you coordinated with @reblace about this refactor? I believe he's aggressively rearranging a bunch modules to support lazy loading. Not sure how far along he is, but it will likely have a bunch of conflicts with this PR.

**Comment by Sieabah** [comment]
Anything dealing with

- User
- UserService
- UserStateService
- AuthenticationService

Ends up having a cyclic dependency at some point which causes undefined classes. No idea why it isn't caught by TypeScript.

An example is user.class.ts which relies on Credentials. I should be able to import from app/authentication but when I try it will export all classes in authentication/*, which User is referenced a lot. 

I believe a lot of the headaches will go away if we split out the classes or try to rely on not having two layers for the application. The UserService should care about users, a TeamService should care about the UserService not TeamService directly interacting with Users, the types should be implied when dealing with specific functions on the UserService.

@ymao1 Yes I was talking to him over the weekend about this.

**Comment by Sieabah** [comment]
I'm pretty sure I've converted all of the files that I could. Anything not using an index.ts has a cyclic dependency that isn't easily resolved with the DI system.

File: src/client/app/admin/authentication/authentication.service.ts [hunk 1]
```diff
@@ -50,7 +51,7 @@ export class AuthenticationService {
 	};
 
 	// Get the user who is currently logged in (or null if no one is logged in)
-	public getCurrentUser(): User {
+	public getCurrentUser() {
```
Comment (andrewmarcus) [inline src/client/app/admin/authentication/authentication.service.ts]: Why remove the return type?

File: src/client/app/admin/authentication/user-state.service.ts [hunk 2]
```diff
@@ -41,4 +45,11 @@ export class UserStateService {
 	public clearUser() {
 		this.user.clearUser();
 	}
+
+	private routeError(err: any): void {
+		// console.log('RouteError');
+		if (err.status === 401) {
+			this.clearUser();
```
Comment (andrewmarcus) [inline src/client/app/admin/authentication/user-state.service.ts:52]: So this is logging out a user entirely anytime they hit any unauthorized page?  Seems a bit draconian...

File: src/client/app/audit/audit.component.ts [hunk 3]
```diff
@@ -7,19 +7,20 @@ import { Observable } from 'rxjs/Observable';
...
-import { TableSortOptions } from '../shared/pageable-table/pageable-table.component';
-import { SortDisplayOption, SortDirection } from '../shared/result-utils.class';
-import { UserService } from '../admin/users.service';
 import { AuditOption } from './audit.classes';
 
 
 @Component({
 	selector: 'audit',
-	templateUrl: './audit-list.component.html'
+	templateUrl: './audit-list.component.html',
```
Comment (andrewmarcus) [inline src/client/app/audit/audit.component.ts]: isn't this trailing comma a JS error?   or does TS suppress it?

File: src/client/app/shared/alert.class.ts [hunk 4]
```diff
@@ -4,3 +4,8 @@ export class Alert {
 		public type: string,
 		public msg: string) {}
 }
+
+export class Alerts {
+	list: Alert[] = [];
+	map: Map<number, Alert> = new Map<number, Alert>();
```
Comment (andrewmarcus) [inline src/client/app/shared/alert.class.ts:10]: I know this wasn't in the original that you took this from, but seems like we'd want add and remove functions that add an alert to both the list and the map.  That can be separate PR though

**Review by andrewmarcus — APPROVED** [review]
I like the direction of this, would be nice to get the User/EUA stuff all worked in too so we could remove more relative paths.

File: src/client/app/audit/audit.component.ts (hunk 3 above)
Comment (Sieabah) [reply src/client/app/audit/audit.component.ts]: It's so you can add/remove lines from an object and you're only editing one line.

Spec: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas

File: src/client/app/admin/authentication/user-state.service.ts (hunk 2 above)
Comment (Sieabah) [reply src/client/app/admin/authentication/user-state.service.ts:52]: That's the current implementation.

File: src/client/app/admin/authentication/authentication.service.ts (hunk 1 above)
Comment (Sieabah) [reply src/client/app/admin/authentication/authentication.service.ts]: Issues with having User be required. It's probably fine now to add it back in but anything to do with AuthenticationService has been pretty painful to load elsewhere.

File: src/client/app/audit/audit.component.ts (hunk 3 above)
Comment (andrewmarcus) [reply src/client/app/audit/audit.component.ts]: Yes, I agree it's more useful to have them, I just didn't realize Javascript let you do that.  Usually I get errors for such things.

Comment (Sieabah) [reply src/client/app/audit/audit.component.ts]: It was adopted in ES5.1, and it doesn't work in function parameters. There is a [proposal](https://github.com/tc39/proposal-trailing-function-commas) for it though.

File: src/client/app/admin/authentication/authentication.service.ts (hunk 1 above)
Comment (reblace) [reply src/client/app/admin/authentication/authentication.service.ts]: Authentication service should be able to depend on user. So, it'd be better to leave the return type here and refactor the code to allow the dependency. We shouldn't be making the code less correct as part of a convenience refactor.

File: config/build/webpack.conf.js [hunk 5]
```diff
@@ -6,6 +6,7 @@ const
 	webpack = require('webpack'),
 	StatsWriterPlugin = require('webpack-stats-plugin').StatsWriterPlugin,
 
+	clientPath = path.posix.resolve('./src/client/'),
```
Comment (reblace) [inline config/build/webpack.conf.js]: If we keep this extracted into a reused variable, please don't list it in with the other require statements.  Just put it in its own block of declared common const variables and maybe name it 'clientSrcPath' to be more descriptive

File: config/build/webpack.conf.js [hunk 6]
```diff
@@ -56,13 +57,13 @@ module.exports = (mode) => {
 	 */
 	if (test) {
 		wpConfig.entry = {
-			application: path.posix.resolve('./src/client/main.ts')
+			application: path.posix.resolve(clientPath+'/main.ts')
```
Comment (reblace) [inline config/build/webpack.conf.js]: Each of these should be something more like:
path.posix.join(clientPath, 'main.ts')

path.posix.resolve() converts a path to an absolute path. So really, we might want to make clientPath = './src/client' and then do the above.

path.posix.join will combine paths (without having to worry about leading/trailing slash wonkiness), which is why I recommend the change

File: config/build/webpack.conf.js [hunk 7]
```diff
@@ -93,6 +94,9 @@ module.exports = (mode) => {
 	 * List of extensions that webpack should try to resolve
 	 */
 	wpConfig.resolve = {
+		alias: {
```
Comment (reblace) [inline config/build/webpack.conf.js:98]: Can you throw a one liner comment above this property that explains what this is achieving?

File: package.json [hunk 8]
```diff
@@ -14,7 +14,9 @@
 		"npm": "3.10.9"
 	},
 
-	"scripts": {},
+	"scripts": {
+		"dev": "export NODE_ENV=development && gulp dev | bunyan"
```
Comment (reblace) [inline package.json]: Can you add build, test, and test-ci while we're at it?

File: src/client/app/admin/authentication/authentication.service.ts (hunk 1 above)
Comment (reblace) [reply src/client/app/admin/authentication/authentication.service.ts]: Dunno if my comment showed up in the other thread, but I'd propose you at least make this return 'any' for now.

File: src/client/app/admin/authentication/inactive.component.ts [hunk 9]
```diff
@@ -1,16 +1,17 @@
...
+	providers: [ User ]
 })
-
 export class InactiveComponent {
-	user: User;
-
 	constructor(
-		private auth: AuthenticationService
+		private auth: AuthenticationService,
+		private user: User
```
Comment (reblace) [inline src/client/app/admin/authentication/inactive.component.ts]: What exactly is the behavior here? Are we injecting a local instance of User into the user property and then overwriting it with the user from the authentication service? Is there a benefit to this approach versus just locally declaring the user variable?

Also, our convention is to mark anything referenced by the template as public. So, user should be public (which it was previously).

File: src/client/app/admin/authentication/no-access.component.ts [hunk 10]
```diff
@@ -1,16 +1,17 @@
...
+	providers: [ User ]
 })
-
 export class NoAccessComponent {
-	user: User;
-
 	constructor(
-		private auth: AuthenticationService
+		private auth: AuthenticationService,
+		private user: User
```
Comment (reblace) [inline src/client/app/admin/authentication/no-access.component.ts]: Same comment as inactive.component.ts

File: src/client/app/admin/authentication/signin.component.ts [hunk 11]
```diff
@@ -1,21 +1,24 @@
...
-	user: User;
 	error: string;
 	config: any;
 
 	constructor(
 		private authService: AuthenticationService,
 		private userStateService: UserStateService,
-		private configService: ConfigService
+		private configService: ConfigService,
+		private user: User
```
Comment (reblace) [inline src/client/app/admin/authentication/signin.component.ts]: Same comment as inactive.component.ts

File: src/client/app/admin/authentication/unauthorized.component.ts [hunk 12]
```diff
@@ -1,16 +1,14 @@
...
-	templateUrl: './unauthorized.component.html'
+	templateUrl: './unauthorized.component.html',
+	providers: [ User ]
 })
-
 export class UnauthorizedComponent {
-	user: User;
-
 	constructor(
+		private user: User
```
Comment (reblace) [inline src/client/app/admin/authentication/unauthorized.component.ts]: Same comment as inactive.component.ts

File: src/client/app/admin/authentication/user-state.service.ts (hunk 2 above)
Comment (reblace) [reply src/client/app/admin/authentication/user-state.service.ts:52]: What's happening here is if the server responds with a 401, it assumes the user has been logged out remotely (or the session has expired), and it will clear the user and go back to the login page (or at least that's what's supposed to happen).

If the user hits a client route they aren't supposed to, they get redirected to the not authorized client page.

File: src/client/app/audit/audit.component.ts (hunk 3 above)
Comment (reblace) [inline src/client/app/audit/audit.component.ts]: We haven't adopted this convention and I'd prefer to not do multiple things in a single branch.

File: src/client/app/admin/password/forgot-password.component.ts [hunk 13]
```diff
@@ -22,7 +22,8 @@ export class ForgotPasswordComponent {
 		private router: Router,
 		private authService: AuthenticationService,
 		private userStateService: UserStateService,
-		private configService: ConfigService
+		private configService: ConfigService,
+		private user: User
```
Comment (reblace) [inline src/client/app/admin/password/forgot-password.component.ts:26]: Same comment as inactive.component.ts

File: src/client/app/admin/user-management/admin-user.module.ts [hunk 14]
```diff
@@ -5,23 +5,23 @@ import { RouterModule } from '@angular/router';
...
-import { UtilModule } from '../../shared/util.module';
-import { AlertService } from '../../shared/alert.service';
-import { CacheEntriesService } from '../../access-checker/cache-entries.service';
-import { UserAudit } from './audit/user-audit.component';
-import { TeamsModule } from '../../teams/teams.module';
 import { UserSignedUpComponent } from './user-signed-up.component';
 
 @NgModule({
-	imports: [
+imports: [
```
Comment (reblace) [inline src/client/app/admin/user-management/admin-user.module.ts]: I think this was unintentional.

File: src/client/app/admin/user/user.class.ts [hunk 15]
```diff
@@ -1,12 +1,9 @@
...
+import { Credentials } from '../authentication/credentials.class';
 
 export class User {
-
-	constructor(
-		public userModel: any = {}, // raw User model object,
-		public credentials: Credentials = new Credentials(),
-		public eua?: any) {
-	}
+	public userModel: any = {}; // raw User model object,
```
Comment (reblace) [inline src/client/app/admin/user/user.class.ts]: the convention we've generally been following is you don't need to explicitly use 'public' since it's the default

File: src/client/app/app.module.ts [hunk 16]
```diff
@@ -1,39 +1,60 @@
+/**
```
Comment (reblace) [inline src/client/app/app.module.ts:1]: I dig the organization here.

File: src/client/app/audit/audit.component.ts (hunk 3 above)
Comment (reblace) [reply src/client/app/audit/audit.component.ts]: I don't have a problem with discussing if we want to start using trailing commas, but it should be a discussion. We haven't enforced it, but we had a convention of not using them. In the interim I'd rather avoid having mixed use. In this case can we revert to not using it (particularly since this change is not directly related to the purpose of this branch/PR)?

File: src/client/app/audit/audit.module.ts [hunk 17]
```diff
@@ -5,15 +5,17 @@ import { NgModule } from '@angular/core';
...
+import { UtilModule } from 'app/shared';
+import { UserService } from 'app/admin/user';
+
 import { AuditViewChangeModal, AuditViewDetailModal } from './audit-view-change.component';
 import { AuditService } from './audit.service';
 import { AuditComponent } from './audit.component';
 import {
-	AuditObjectComponent, UrlAudit,
-	DefaultAudit, ExportAudit
+AuditObjectComponent, UrlAudit,
```
Comment (reblace) [inline src/client/app/audit/audit.module.ts]: Indenting

File: tsconfig.json [hunk 18]
```diff
@@ -19,12 +19,12 @@
 		"noUnusedParameters": false,
 		"noUnusedLocals": false,
 
-		"sourceMap": true
+		"sourceMap": true,
 
-	},
-
-	"files": [
```
Comment (reblace) [inline tsconfig.json:26]: You need to maintain the files list here for IntelliJ and presumably other tools to work properly.

**Review by reblace — CHANGES_REQUESTED** [review]
There's a bunch of comments throughout. Before we merge, I'd like to understand exactly why cyclical dependencies are only an issue after this change (ie. is it the fact that we're rolling things up into index.ts modules? or the fact that we're aliasing the app path?) Cause, it seems like we're creating a lot of pain when cyclical dependencies were never an issue before. Presumably the Typescript compiler (like other compilers) is able to deal with cyclical dependencies.

**Comment by Sieabah** [comment]
@reblace So Typescript is able to determine the cyclic dependencies fine. It's when the decorators are evaluated that causes the problem. The Angular DI system when the decorators are run cause some to be undefined due to the cyclic dependencies. It's an issue when using index.ts as every export statement is expected to finish and export (for Angular DI).

The solution is to really start to decouple the classes that have this problem.

File: src/client/app/admin/user-management/admin-user.module.ts (hunk 14 above)
Comment (Sieabah) [reply src/client/app/admin/user-management/admin-user.module.ts]: Yep.

File: src/client/app/admin/password/forgot-password.component.ts (hunk 13 above)
Comment (Sieabah) [reply src/client/app/admin/password/forgot-password.component.ts:26]: Do you want me to remove it then? I don't know if any projects rely on it.

File: tsconfig.json (hunk 18 above)
Comment (Sieabah) [reply tsconfig.json:26]: It still works fine for my IntelliJ. The baseDir is what it goes off of.

File: src/client/app/admin/authentication/inactive.component.ts (hunk 9 above)
Comment (Sieabah) [reply src/client/app/admin/authentication/inactive.component.ts]: This was originally when I was trying to figure out why User/AuthenticationService was undefined during runtime.

File: src/client/app/audit/audit.module.ts (hunk 17 above)
Comment (Sieabah) [reply src/client/app/audit/audit.module.ts]: Probably happened when I was reordering lines, IntelliJ likes to remove the indents.

File: src/client/app/admin/password/forgot-password.component.ts (hunk 13 above)
Comment (Sieabah) [reply src/client/app/admin/password/forgot-password.component.ts:26]: Didn't realize the context of the statement.

**Comment by Sieabah** [comment]
@reblace To further clarify why load-order is suddenly making a difference. When loading via index.ts it exports everything before actually resolving to the parent index.ts or whatever module is requesting it. In doing so the dependencies are deferred to resolve later (cyclic). Typescript doesn't complain and neither would javascript. The issue is the decorators needing the necessary metadata at compile time which when they're executed the actual definition is deferred. So we get a nice big undefined or ? warning from the Angular DI system at runtime.

Edit: This wouldn't change tree shaking as it still breaks out what isn't used from the exports.

File: src/client/app/admin/authentication/user-state.service.ts (hunk 2 above)
Comment (brianghig) [reply src/client/app/admin/authentication/user-state.service.ts:52]: I think the difference is between 401 (which is supposed to be for authentication) vs. 403 (which is supposed to be used for authorization). 403s should be used for unauthorized resources, whereas 401s would be returned if the user isn't authenticated or authenticated improperly (e.g., lost session). http://stackoverflow.com/a/6937030

File: tsconfig.json (hunk 18 above)
Comment (reblace) [reply tsconfig.json:26]: Oh ok, I can try out the baseDir and see if IntelliJ stops freaking out. I've tried a lot of different permutations over time to keep the build happy and keep IntelliJ happy

Comment (Sieabah) [reply tsconfig.json:26]: As long as you have the tsconfig in the root of the project/module I've had no problems with it finding modules.

File: src/client/app/shared/alert.class.ts (hunk 4 above)
Comment (Sieabah) [reply src/client/app/shared/alert.class.ts:10]: It was in the original: https://github.com/Asymmetrik/mean2-starter/pull/118/files#diff-0c5068ddc631b166e116c9b74fe82d13L10

File: tsconfig.json (hunk 18 above)
Comment (reblace) [reply tsconfig.json:26]: What I've found is that IntelliJ will find the tsconfig if it's in the root path, but it won't actually apply all of the settings to your ts files if you haven't identified (either using "files" or "include") which files it should apply to. There are even situations where tsc will complain if both the "files" or "include" list are empty. One big example that I keep running into is the experimental decorators support (since that isn't on be default).

Comment (Sieabah) [reply tsconfig.json:26]: No idea what I have configured differently then because all the settings are applied.

**Comment by Sieabah** [comment]
@reblace Was there anything else you wanted to add?

**Comment by Sieabah** [comment]
Changes Made So Far:
- [Change from AsyHttp interfacing to User and instead have User subscribe to AsyHttp errors](https://github.com/Asymmetrik/mean2-starter/pull/118/files#diff-cd3c262c75147f5c26404e307d5432a5R46)
- [Add index.ts to every folder which exports subfolder and same-directory classes](https://github.com/Asymmetrik/mean2-starter/pull/118/files#diff-df29983be0be56373814b5d8b237d36fR1)
- [Update imports to use package paths from root instead of using hard relative paths](https://github.com/Asymmetrik/mean2-starter/pull/118/files#diff-bc37603b6132820ebf1270229340d357R3)

**Comment by Sieabah** [comment]
How to fix Cyclic Dependencies

Manage state through a shared top level service. Components having interdependencies creates a layer that is solved during compilation but after decorators are called.

This is only an issue when you use index.ts as every file referenced in the index.ts file you reference must finish exporting. 

So for example if we have folders, A and B. A has an index which exports a1 and a2. B exports b1.
If b1 imports a1:  _import { a1 } from '../A'; _ and a1 imports b1: _import {b1} from '../B'_; In Typescript the compiler as it's reading b1 will wait until a1 is defined, and will take into account the cyclic dependency and resolve.

The issue is decorators which are abundant in Angular2 are [executed at compile time](https://www.typescriptlang.org/docs/handbook/decorators.html). They are also still _experimental_. So during the time where there is a cyclic dependency (even if indirect) the decorator will be executed and the metadata that Angular uses is unavailable which results in the DI system having the reference defined and injectable but the object injected is undefined or null.

An example of a highly cyclic dependency is AsyHttp and the User class. Before it was a direct link from AsyHttp (which a lot of components/services used) to User. Meaning almost any class had to have User exported, and since AsyHttp would have exported through 'admin/user' or 'admin', any service or module in the exported in the relevant export would also have to be exported. So by including AsyHttp in something like Teams your dependency chain results in Team -> AsyHttp -> 'admin' (index.ts) -> (**User**, ..., user-management) -> (..., admin-list-users.component) -> 'teams' (index.ts) -> (Team, ...). This chain results in loading almost the entire application.

So in short, Typescript doesn't have a problem, but AngularDI does. Using index.ts takes out the hard requirement of importing specific files, any interaction with services should be done through AngularDI. State should be handled with something like Redux or a giant object that is injected.

**Comment by Sieabah** [comment]
Why to use root alias:

So the root alias for 'app/*' that I added is to simplify importing from outside your current module. The issue without using paths is that you lock yourself into strict _this file is this many folders deep_. It makes reorganization later a real hassle. Take for example when you require from npm, you don't do nest your dependencies down to the implementation `require('../../../node_modules/lodash/get/index.js');` You just simply do `require('lodash')` and rely on npm to assume since it didn't start with `.` you meant the node_modules folder. Also from that you expect the exported object to have what you need where you need it.

If I know I need the User class in the admin module and I'm in the teams module I can simply rely on the exported module 'app/admin' to have what I need and I don't need to worry about the underlying file structure of that module. Since it is an alias you are still free to do app/admin/user/user.class and pull in your exported user, but if you use it in conjunction with index.ts you get a very nicely packaged module layout.

So aliasing `app/*` I believe is unique enough, it can be whatever you want asy_app, appASY, LOCAL_APPLICATION_FOLDER_ASYMMETRIK, literally anything you want. If you're worried about accidentally pulling in [app](https://www.npmjs.com/package/app) as a dependency and meaning to use that, then yes, use a different folder name. It can be named wildfire/firehawk. It can even be named app in your directory and aliased in tsconfig as apollo/*.

**Comment by Sieabah** [comment]
Bump
